servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherCache.java

```diff
@@ -0,0 +1,261 @@
+                            public void onNext(@Nullable T next) {
+                                subscriber.onNext(next);
+                            }
+
+                            @Override
+                            public void onError(Throwable t) {
+                                lockRemoveFromMap();
+                                subscriber.onError(t);
+                            }
+
+                            @Override
+                            public void onComplete() {
+                                lockRemoveFromMap();
+                                subscriber.onComplete();
+                            }
+
+                            private void lockRemoveFromMap() {
+                                synchronized (publisherCache) {
+                                    publisherCache.remove(key, item2);
+                                }
+                            }
+                        });
+
+                item2.publisher = multicastStrategy.apply(newPublisher)
+                        .liftSync(subscriber -> new Subscriber<T>() {
+                            @Override
+                            public void onSubscribe(Subscription subscription) {
+                                subscriber.onSubscribe(new Subscription() {
+                                    @Override
+                                    public void request(long n) {
+                                        subscription.request(n);
+                                    }
+
+                                    @Override
+                                    public void cancel() {
+                                        synchronized (publisherCache) {
+                                            subscription.cancel();
+                                        }
+                                    }
+                                });
+                            }
+
+                            @Override
+                            public void onNext(@Nullable T next) {
+                                subscriber.onNext(next);
+                            }
+
+                            @Override
+                            public void onError(Throwable t) {
+                                try {
+                                    subscriber.onError(t);
+                                } finally {
+                                    lockRemoveFromMap();
+                                }
+                            }
+
+                            @Override
+                            public void onComplete() {
+                                try {
+                                    subscriber.onComplete();
+                                } finally {
+                                    lockRemoveFromMap();
+                                }
+                            }
+
+                            private void lockRemoveFromMap() {
+                                synchronized (publisherCache) {
```

Comment:
add a comment here to the effect:

- completion of the first Subscriber after multicast means the the multicast operator is in a terminal state and we  therefore remove it from the map. There are cases where folks may want to re-subscribe to the Publisher (e.g. get the cached value, trigger another event) however that currently isn't supported and we favor bounding the size of the map which has scope outside the operator chain.